src/core/backend_context.cc

```diff
@@ -291,24 +295,55 @@ BackendResponder::Finalize()
                 status),
             "error sending TensorFlow response");
       } else {
-        bool cuda_used = false;
-        status = CopyBuffer(
-            response_output->Name(), pinned_memory_type, pinned_memory_id,
-            response_memory_type, response_memory_type_id, response_byte_size,
-            pinned_buffer + offset, const_cast<void*>(response_buffer), stream_,
-            &cuda_used);
-        need_sync_ |= cuda_used;
+        // Only use parallel CopyBuffer for CPU-CPU copy when worker count > 1
+        if ((AsyncWorkQueue::GetWorkerCount() > 1) &&
+            (response_memory_type == TRITONSERVER_MEMORY_CPU) &&
+            (pinned_memory_type == TRITONSERVER_MEMORY_CPU)) {
+          AsyncWorkQueue::AddTask(std::bind(
+              CopyBufferHandler, response_output->Name(), pinned_memory_type,
+              pinned_memory_id, response_memory_type, response_memory_type_id,
+              response_byte_size, pinned_buffer + offset,
+              const_cast<void*>(response_buffer), stream_,
+              &cuda_used_vec[count], &completion_queue));
+          count++;
+        } else {
+          status = CopyBuffer(
+              response_output->Name(), pinned_memory_type, pinned_memory_id,
+              response_memory_type, response_memory_type_id, response_byte_size,
+              pinned_buffer + offset, const_cast<void*>(response_buffer),
+              stream_, &cuda_used_vec[count]);
 
-        if (!status.IsOk()) {
+          if (!status.IsOk()) {
+            LOG_STATUS_ERROR(
+                InferenceResponse::SendWithStatus(
+                    std::move(*response), TRITONSERVER_RESPONSE_COMPLETE_FINAL,
+                    status),
+                "error sending TensorFlow response");
+          }
+        }
+      }
+
+      offset += response_byte_size;
+    }
+
+    for (size_t i = 0; i < cuda_used_vec.size(); i++) {
```

Comment:
You can't do it this way because you have no way of knowing that the task have yet set cuda_used_vec. Put cuda_used values in the completion queue with the status as we discussed.

Reply:
Done